### PR TITLE
fix: update form input validator to remove errors if appropriate on re-validation

### PR DESCRIPTION
**Pull request by crespire**
Complete the following REQUIRED checkboxes:

-   [x] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/theodinproject/blob/main/CONTRIBUTING.md)
-   [x] The title of this PR follows the `keyword: brief description of change` format, using one of the following keywords:
  - `Feature` - adds new or amends existing user-facing behaviour
  - `Chore` - changes that have no user-facing value, refactors, dependency bumps, etc
  - `Fix` - bug fixes

Complete the following checkbox ONLY IF it is applicable to your PR. You can complete it later if it is not currently applicable:
-   [x] I have verified all tests and linters pass against my changes, and/or I have included automated tests where applicable

<hr>

**1. Because:**
Form validation controller was not removing error messages on re-validation after identifying an error which is then corrected by the user.

**2. This PR:**
* Updates `form_validation_controller.js` to remove form errors, and re-adjust input style after validation if there are no errors present in the field.
* Adds re-validation tests to confirm behavior to the following spec files:
  * `sign_up_spec.rb`
  * `user_login_spec.rb`
  * `user_settings_spec.rb`

**3. Additional Information:**

Thread here: https://discord.com/channels/505093832157691914/1026938313610907779

### CODE REVIEW

File: app/javascript/controllers/form_validation_controller.js [hunk 1]
```diff
@@ -61,6 +63,26 @@ export default class FormValidation extends ValidationController {
...
     );
   }
+
+  static removeError(el) {
+    const errorField = el.parentElement.nextElementSibling;
+    errorField.classList.add('hidden');
+    errorField.textContent = '';
+    el.classList.remove(...el.classList);
+    el.classList.add(
+      'block',
```
Comment (KevinMulhern) [inline app/javascript/controllers/form_validation_controller.js:73]: Having the default form styles duplicated here introduces a little bit of brittleness, we'd need to remember to update here and the builder if we ever need to update the form field designs. 

We could alternatively take the invalid field classes on [line 59](https://github.com/TheOdinProject/theodinproject/blob/54d17a8b714274c64951ceb2ababf45f7890ed2d/app/javascript/controllers/form_validation_controller.js#L59) and put them into a const above the class definition:

```javascript
const invalidFieldClasses = [
  'border-red-300',
  'text-red-900',
  'placeholder-red-300',
  'focus:ring-red-500',
  'focus:border-red-500',
]

export default class FormValidation extends ValidationController {
  ...
}
```

Then we can simplify adding and removing error styles
```javascript
static applyErrorStylesTo(el) {
    el.classList.add(...invalidFieldClasses);
  }

  static removeError(el) {
    const errorField = el.parentElement.nextElementSibling;
    errorField.classList.add('hidden');
    errorField.textContent = '';
    el.classList.remove(...invalidFieldClasses);
  }

```

**Review by KevinMulhern — CHANGES_REQUESTED** [review]
This works a treat, great stuff @crespire 💪  
I Just a single suggestion for feedback.

File: app/javascript/controllers/form_validation_controller.js (hunk 1 above)
Comment (crespire) [reply app/javascript/controllers/form_validation_controller.js]: Fixed with [13a3cd1](https://github.com/TheOdinProject/theodinproject/pull/3380/commits/13a3cd1e5295b1506531b7cf5635014c7e0c5561)!

File: app/javascript/controllers/form_validation_controller.js [hunk 2]
```diff
@@ -38,12 +46,14 @@ export default class FormValidation extends ValidationController {
...
       FormValidation.errorMessageEl(el).textContent = this.errorMessage(attr);
+    } else {
+      FormValidation.removeError(el);
     }
   }
 
   static errorMessageEl(el) {
     const errorField = el.parentElement.nextElementSibling;
-    errorField.classList.remove('hidden');
+    errorField.classList.toggle('hidden');
```
Comment (KevinMulhern) [inline app/javascript/controllers/form_validation_controller.js:56]: Was this change intentional @crespire?

Whiling QA'ing, I noticed the some odd behaviour due to the toggling:

https://user-images.githubusercontent.com/7963776/194119534-69f332c0-6b4c-4562-8aec-83a2ee7e2800.mov

**Review by KevinMulhern — CHANGES_REQUESTED** [review]

File: app/javascript/controllers/form_validation_controller.js (hunk 2 above)
Comment (KevinMulhern) [reply app/javascript/controllers/form_validation_controller.js:56]: But it does work as expected when we use `remove` here instead as far as I can tell.

Comment (crespire) [reply app/javascript/controllers/form_validation_controller.js:56]: Ah, in that case, I can change it back! I thought toggle might be a better approach. I did some light QA and ran the tests after changing to `toggle` but seems `add/remove` is a better approach.

Will revert that commit now.

**Review by KevinMulhern — APPROVED** [review]
Great work @crespire! 🏆
